rosbag2_cpp/include/rosbag2_cpp/writers/buffer_layer.hpp

```diff
@@ -0,0 +1,131 @@
+// Copyright 2020, Robotec.ai sp. z o.o.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef ROSBAG2_CPP__WRITERS__BUFFER_LAYER_HPP_
+#define ROSBAG2_CPP__WRITERS__BUFFER_LAYER_HPP_
+#include <atomic>
+#include <condition_variable>
+#include <memory>
+#include <mutex>
+#include <thread>
+#include <vector>
+
+#include "rosbag2_storage/storage_interfaces/read_write_interface.hpp"
+#include "rosbag2_storage/storage_options.hpp"
+#include "rosbag2_cpp/visibility_control.hpp"
+
+namespace rosbag2_cpp
+{
+namespace writers
+{
+
+class ROSBAG2_CPP_PUBLIC BufferLayer
+{
+public:
+  BufferLayer(
+    std::shared_ptr<rosbag2_storage::storage_interfaces::ReadWriteInterface> storage,
```

Comment:
somewhat related to the comment with the threading, but I would appreciate if we could separate the storage from the buffer. I feel that should be a separation of concerns here. The buffer really only buffers incoming data and should not have a (strong) dependency to the storage backend.
Could we maybe add an API which takes a lamdba function or some sort of callback which is getting called with a filled buffer to be written?

The reason why I am arguing for this is have introduce a bit of flexibility on what to do with this buffer or have a simply way to bypass this in total.